### PR TITLE
add conv higher order derivative

### PR DESCRIPTION
**Pull request by reygu**
- 为 conv 添加二阶导算子
- 问题来源： https://github.com/Oneflow-Inc/OneCloud/issues/106 
- 因输入 x 是 ConvFilterGrad 的输入，所以在 ConvFilterGradGrad 中计算 x 的二阶导，同理，在 ConvDataGradGrad 中计算权重 w 的二阶导

### CODE REVIEW

File: oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp [hunk 1]
```diff
@@ -167,5 +168,185 @@ class LeakyReluGradGrad : public OpExprGradFunction<LeakyReluGradGradCaptureStat
...
+  ctx->kernel_size = JUST(composed_attrs.GetAttr<std::vector<int32_t>>("kernel_size"));
+  ctx->strides = JUST(composed_attrs.GetAttr<std::vector<int32_t>>("strides"));
+  ctx->dilation_rate = JUST(composed_attrs.GetAttr<std::vector<int32_t>>("dilation_rate"));
+  ctx->groups = JUST(composed_attrs.GetAttr<int32_t>("groups"));
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> ConvDataGradGrad::Apply(const ConvDataGradGradCaptureState* ctx,
+                                    const TensorTuple& out_grads, TensorTuple* in_grads) const {
+  in_grads->resize(2);
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp:238]: 这里的 size 应该是 3？

File: oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp [hunk 2]
```diff
@@ -167,5 +168,185 @@ class LeakyReluGradGrad : public OpExprGradFunction<LeakyReluGradGradCaptureStat
...
+  size_t num_spatial_dims = ctx->kernel_size.size();
+
+  // w_grad_grad = out_grads_x * grad
+  if (ctx->w_requires_grad) {
+    const auto& grad = ctx->SavedTensors().at(ctx->grad_index);
+    in_grads->at(0) = JUST(functional::ConvFilterGrad(
+        grad, out_grads.at(0), num_spatial_dims, ctx->kernel_size, ctx->strides,
+        ctx->padding_before, ctx->dilation_rate, ctx->groups, ctx->data_format));
+  }
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp:247]:
```cpp
    op_ = CHECK_JUST(one::OpBuilder("conv_data_grad")
                         .Input("dy")
                         .Input("filter")
                         .Input("x_like")
                         .Output("dx")
                         .Build());
```

这里需要观察 ConvDataGradFunctor 内声明的 input 的顺序，可以发现 w_grad_grad 是在下标为 1 的位置，这里赋值就赋错位了

Comment (REYGU) [reply oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp:247]: 好的，之前不知道跟这里是对应的

File: oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp (hunk 1 above)
Comment (REYGU) [reply oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp:238]: 这个地方实际只求两个，x_like 是不求的，但是还是需要预留 x_like 的位置吗

Comment (wyg1997) [reply oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp:238]: 对的，不求的话留一个位置，但是值为空就行

Comment (REYGU) [reply oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp:238]: OK

Comment (wyg1997) [reply oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp:238]: 那边调用的地方对 x_like 要 detach 一下别忘了

File: python/oneflow/test/modules/test_global_higher_derivative_conv.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+import numpy as np
+import oneflow as flow
+import oneflow.unittest
+from oneflow.test_utils.automated_test_util import *
+
+import torch as pytorch_origin
+import oneflow as oneflow_origin
+
+
+def _test_convnd_grad_grad_impl(test_case, ndim, placement, x_sbp, w_sbp):
```
Comment (REYGU) [inline python/oneflow/test/modules/test_global_higher_derivative_conv.py:27]: globaltest 这里有一些疑惑，参考的 matmul/deconv 的 globaltest，x/w 设置了不同的 sbp，这里涉及到的变量有 `x, w, y, init_grad_x, init_grad_w, init_grad_y` 对应的 sbp 应如何设置？

File: oneflow/core/autograd/gradient_funcs/conv.cpp [hunk 4]
```diff
@@ -85,8 +85,8 @@ Maybe<void> ConvolutionNd::Apply(const ConvolutionNdCaptureState* ctx, const Ten
     const auto& weight = ctx->SavedTensors().at(ctx->weight_index);
     const auto& input = ctx->SavedTensors().at(ctx->input_index);
     in_grads->at(0) = JUST(functional::ConvDataGrad(
-        out_grads.at(0), weight, input, num_spatial_dims, ctx->kernel_size, ctx->strides,
-        ctx->padding_before, ctx->dilation_rate, ctx->groups, ctx->data_format));
+        out_grads.at(0), weight, JUST(input->detach()), num_spatial_dims, ctx->kernel_size,
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/conv.cpp:88]: 哦这里你理解错了，应该在 ConvDataGrad 的 functor 内做 detach，保证以后每一次调用 ConvDataGrad 都是正确捕获的，在这里加的话，就把责任丢给调用者了，这是不对的

File: python/oneflow/test/modules/test_global_higher_derivative_conv.py (hunk 3 above)
Comment (wyg1997) [inline python/oneflow/test/modules/test_global_higher_derivative_conv.py:27]: 可以这样，不传 x 和 w 的 sbp，下面用 random_sbp 来生成，然后你本地循环 1000 次来测试，提交给 CI 的时候跑 5 次 10 次就行（看看这个单测的时间，控制在 30s 以下就行）

File: oneflow/core/autograd/gradient_funcs/conv.cpp (hunk 4 above)
Comment (REYGU) [reply oneflow/core/autograd/gradient_funcs/conv.cpp]: 已修改

File: python/oneflow/test/modules/test_global_higher_derivative_conv.py (hunk 3 above)
Comment (REYGU) [reply python/oneflow/test/modules/test_global_higher_derivative_conv.py]: 已修改，发现不是 sbp 的原因... 把 np 初始化的 tensor 都改成 random_tensor 就可以了

Comment (wyg1997) [reply python/oneflow/test/modules/test_global_higher_derivative_conv.py]: 哦哦，这个问题是因为如果用 ndarray 初始化 tensor 的话，在 to_global 的时候得到的 shape 可能不是预期那样，比如 `flow.tensor(np.random.rand((8, 8))).to_global(placement, sbp=flow.sbp.split(0))` ，得到的结果可能是 (16, 8) 的 tensor，就会报 shape 对不上的问题了

File: python/oneflow/test/modules/test_global_higher_derivative_conv.py [hunk 5]
```diff
@@ -24,34 +24,25 @@
...
-        .requires_grad_(True)
-        .to_global(placement=placement, sbp=x_sbp)
+def _test_convnd_grad_grad_impl(test_case, ndim, placement, x_sbp):
+    x_shape = [8, 8] + [5 for _ in range(ndim)]
+    w_shape = [8, 8] + [3 for _ in range(ndim)]
+    y_shape = [8, 8] + [3 for _ in range(ndim)]
+
+    x = random_tensor(len(x_shape), *x_shape).to_global(
+        placement=placement, sbp=x_sbp
     )
```
[the remark this answers is not in the document]
Comment (REYGU) [reply python/oneflow/test/modules/test_global_higher_derivative_conv.py]: 已修改

File: oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp [hunk 6]
```diff
@@ -167,5 +168,200 @@ class LeakyReluGradGrad : public OpExprGradFunction<LeakyReluGradGradCaptureStat
...
+                                      const TensorTuple& outputs, const AttrMap& attrs) const {
+  // input: dy, w, x_like, [add to output]
+  // output: dx
+  CHECK_EQ_OR_RETURN(inputs.size(), 3);   // NOLINT(maybe-need-error-msg)
+  CHECK_EQ_OR_RETURN(outputs.size(), 1);  // NOLINT(maybe-need-error-msg)
+
+  ctx->w_requires_grad = inputs.at(1)->requires_grad();
+  ctx->grad_requires_grad = inputs.at(0)->requires_grad();
+
+  if (!(ctx->w_requires_grad || ctx->grad_requires_grad)) { return Maybe<void>::Ok(); }
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp:215]: 这一行不需要，因为如果这两个输入 requires_grad 都是 False 的话，这个后向就不会构建

File: oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp [hunk 7]
```diff
@@ -167,5 +168,200 @@ class LeakyReluGradGrad : public OpExprGradFunction<LeakyReluGradGradCaptureStat
...
+                                        const AttrMap& attrs) const {
+  // input: dy, x
+  // output: dw
+  CHECK_EQ_OR_RETURN(inputs.size(), 2);   // NOLINT(maybe-need-error-msg)
+  CHECK_EQ_OR_RETURN(outputs.size(), 1);  // NOLINT(maybe-need-error-msg)
+
+  ctx->x_requires_grad = inputs.at(1)->requires_grad();
+  ctx->grad_requires_grad = inputs.at(0)->requires_grad();
+
+  if (!(ctx->x_requires_grad || ctx->grad_requires_grad)) { return Maybe<void>::Ok(); }
```
Comment (wyg1997) [inline oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp:311]: 同上

File: oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp (hunk 6 above)
Comment (REYGU) [reply oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp]: done

File: oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp (hunk 7 above)
Comment (REYGU) [reply oneflow/core/autograd/gradient_funcs/higher_derivative_grad.cpp]: done

File: python/oneflow/test/modules/test_higher_derivative_conv.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        grad_outputs=init_grad_y,
+        create_graph=True,
+        retain_graph=True,
+    )[0]
+    test_case.assertTrue(
+        np.allclose(dw.pytorch.detach().cpu().numpy(), dw.oneflow.detach().numpy())
+    )
+
+
+
```
Comment (wyg1997) [inline python/oneflow/test/modules/test_higher_derivative_conv.py:72]: 这里留一个空行就可以，函数中的逻辑不需要这种大段的空行

**Review by wyg1997 — APPROVED** [review]
LGTM 👍

File: python/oneflow/test/modules/test_higher_derivative_conv.py (hunk 8 above)
Comment (REYGU) [reply python/oneflow/test/modules/test_higher_derivative_conv.py]: 已修改

**Review by hjchen2 — APPROVED** [review]

File: oneflow/core/autograd/higher_order_gradient_funcs/conv.cpp [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+                                    const TensorTuple& out_grads, TensorTuple* in_grads) const {
+  in_grads->resize(3);
+  size_t num_spatial_dims = ctx->kernel_size.size();
+
+  // x_grad       = grad * w            (y.shape * w.shape -> x.shape)  call ConvDataGrad
+  // x_grad_grad  = grad * out_grads_w  (y.shape * w.shape -> x.shape)  call ConvDataGrad
+  // grad_grad_x  = out_grads_x * w     (x.shape * w.shape -> y.shape)  call ConvND
+  // w_grad       = x * grad            (x.shape * y.shape -> w.shape)  call ConvFilterGrad
+  // w_grad_grad  = out_grads_x * grad  (x.shape * y.shape -> w.shape)  call ConvFilterGrad
+  // grad_grad_w  = x * out_grads_w     (x.shape * w.shape -> y.shape)  call ConvND
```
Comment (Ldpe2G) [inline oneflow/core/autograd/higher_order_gradient_funcs/conv.cpp:93]: 我觉得注释要写的更加详细易懂一些，要结合下面的代码才能看懂注释和注释中的符号指代什么。
可以先从 一阶导开始解释，
比如  `w * x = y`，
则梯度 `w_grad = y_grad * x.t`， `x_grad = w.t * y_grad`，

然后就是，二阶导 
二阶前向（一阶反向）：`w_grad = y_grad * x.t`
二阶反向：  ` grad_for_y_grad = out_grads_for_w_grad * x`， `grad_for_x.t = y_grad.t * out_grads_for_w_grad`

二阶前向（一阶反向）：`x_grad = w.t * y_grad`
二阶反向：  ` grad_for_y_grad = w * out_grads_for_x_grad`， `grad_for_w.t = out_grads_for_x_grad * y_grad.t`

Comment (reygu) [reply oneflow/core/autograd/higher_order_gradient_funcs/conv.cpp:93]: 好的，我再调整一下
